File: Ch02/Ex2_06.hs
{-
 - Ex2_06.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Church Numerals
 -
 -}

zero   = const id
add1 n = const ()
one    = id
two f  = f . f

churchEncode n =
  (`innerEncode` n)
  where
    innerEncode f n x
      | n == 0    = x
      | otherwise = f (innerEncode f (n-1) x)

File: Ch02/Ex2_06.scala
/**
  * Ex2_06.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * Church numerals
  */

type F = Any => Any

val zero = (f: F) => ((x: Any) => x)
val add1 = (n: F) => (f: F) => ((x: Any) => (f((n(f)(x))))
val plus = (n: F, m: F) => ((f: F) => ((x: Any) => (m(f)(n(f)(x)))))
val one  = (f: F) => ((x: Any) => f(x))
val two  = (f: F) => ((x: Any) => f(f(x)))

def churchEncode(n: Int) = {
  def innerEncode(f: F, n: Int, x: Any): Any = {
    n match {
      case 0 => x
      case _ => f(innerEncode(f, n, x))
    }
  }
  (f: F) => (x: Any) => innerEncode(f, n, x)
}

def output() = {
  "Not sure how to test this. It compiles with strong types, which isn't *nothing*..."
}

/**
  * execute with
  * scala -i Ex2_06.scala -e "println(output)"
  * until I figure out a better way
  */
